# Francis Yang 10/14/2022
## Pandas Assignment
1. How many rows of data are in the DataFrame?

In [1]:
import pandas as pd
data = pd.read_csv('datasets/ksprojects.csv')
data.shape

(65418, 13)

There are 65,418 rows in this dataframe

2. What are the names and data types of the columns?

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65418 entries, 0 to 65417
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             65418 non-null  int64  
 1   name           65418 non-null  object 
 2   category       65418 non-null  object 
 3   main_category  65418 non-null  object 
 4   deadline       65418 non-null  object 
 5   goal           65418 non-null  float64
 6   launched       65418 non-null  object 
 7   state          65418 non-null  object 
 8   backers        65418 non-null  int64  
 9   country        65418 non-null  object 
 10  usd_pledged    64673 non-null  float64
 11  currency       65418 non-null  object 
 12  pledged        65418 non-null  float64
dtypes: float64(3), int64(2), object(8)
memory usage: 6.5+ MB


3. Do any of the columns contain null values?

In [3]:
nullcheck = data.isnull()
for col in nullcheck.columns:
    print(nullcheck[col].value_counts())
    print()

False    65418
Name: id, dtype: int64

False    65418
Name: name, dtype: int64

False    65418
Name: category, dtype: int64

False    65418
Name: main_category, dtype: int64

False    65418
Name: deadline, dtype: int64

False    65418
Name: goal, dtype: int64

False    65418
Name: launched, dtype: int64

False    65418
Name: state, dtype: int64

False    65418
Name: backers, dtype: int64

False    65418
Name: country, dtype: int64

False    64673
True       745
Name: usd_pledged, dtype: int64

False    65418
Name: currency, dtype: int64

False    65418
Name: pledged, dtype: int64



There are some null values in the `usd_pledged` column

4. Find all successful documentary projects and sort them by the amount pledged. Print the top 10 highest pledges.


In [4]:
data = data.sort_values('pledged',ascending = False)

5. Create a new column named average_per_backer and set its value to the total amount pledged / number of backers.


In [5]:
data['average_per_backer'] = data['pledged']/data['backers']
data.head()

id                                            name  \
12253  1033978702          OUYA: A New Kind of Video Game Console   
62904  1386523707                   Fidget Cube: A Vinyl Desk Toy   
26917  1152735261  Pono Music - Where Your Soul Rediscovers Music   
45290  1074800880               Bloodstained: Ritual of the Night   
46419  1045500705                         Zombicide: Black Plague   

              category main_category         deadline      goal  \
12253  Gaming Hardware         Games    8/9/2012 7:00  950000.0   
62904   Product Design        Design  10/20/2016 3:00   15000.0   
26917            Sound    Technology  4/15/2014 19:30  800000.0   
45290      Video Games         Games   6/13/2015 5:00  500000.0   
46419   Tabletop Games         Games    7/7/2015 2:00  125000.0   

              launched       state  backers country  usd_pledged currency  \
12253  7/10/2012 14:44  successful    63416      US    8600000.0      USD   
62904  8/30/2016 22:02  successful   154926      US      13770.0      USD   
26917  3/11/2014 18:30  successful    18220      US    6230000.0      USD   
45290  5/11/2015 18:58  successful    64867      US    5550000.0      USD   
46419   6/8/2015 20:59  successful    20915      US    4080000.0      USD   

         pledged  average_per_backer  
12253  8600000.0          135.612464  
62904  6470000.0           41.761873  
26917  6230000.0          341.931943  
45290  5550000.0           85.559684  
46419  4080000.0          195.075305

6. What happened to the rows with 0 backers? How can this be dealt with?


This would cause division by zero. We could add a conditional statement or replace values with Nan

7. Drop all rows with 0 backers then repeat the previous exercise.


In [6]:
data['average_per_backer'].isna().value_counts()

False    56136
True      9282
Name: average_per_backer, dtype: int64

In [7]:
data2 = data.dropna(subset=['average_per_backer'])
data2

id                                            name  \
12253  1033978702          OUYA: A New Kind of Video Game Console   
62904  1386523707                   Fidget Cube: A Vinyl Desk Toy   
26917  1152735261  Pono Music - Where Your Soul Rediscovers Music   
45290  1074800880               Bloodstained: Ritual of the Night   
46419  1045500705                         Zombicide: Black Plague   
...           ...                                             ...   
43036  1387088345                    Will and Dan's Adventure Van   
50633  1049445142                                  Alleanza Comic   
63275  1354388458                          Bamzy Baby Productions   
40169  1237363045                                     StudentList   
6198   1368692455                      The 99% - We Will Be Heard   

              category main_category          deadline      goal  \
12253  Gaming Hardware         Games     8/9/2012 7:00  950000.0   
62904   Product Design        Design   10/20/2016 3:00   15000.0   
26917            Sound    Technology   4/15/2014 19:30  800000.0   
45290      Video Games         Games    6/13/2015 5:00  500000.0   
46419   Tabletop Games         Games     7/7/2015 2:00  125000.0   
...                ...           ...               ...       ...   
43036           Places   Photography   4/30/2015 14:21    1248.0   
50633           Comics        Comics  10/21/2015 14:21     300.0   
63275        Animation  Film & Video   11/1/2016 18:11   15000.0   
40169              Web    Technology    3/6/2015 21:47    1000.0   
6198      Film & Video  Film & Video    11/1/2011 5:59    5000.0   

               launched       state  backers country   usd_pledged currency  \
12253   7/10/2012 14:44  successful    63416      US  8.600000e+06      USD   
62904   8/30/2016 22:02  successful   154926      US  1.377000e+04      USD   
26917   3/11/2014 18:30  successful    18220      US  6.230000e+06      USD   
45290   5/11/2015 18:58  successful    64867      US  5.550000e+06      USD   
46419    6/8/2015 20:59  successful    20915      US  4.080000e+06      USD   
...                 ...         ...      ...     ...           ...      ...   
43036   3/31/2015 14:21      failed        1      AU  7.723840e-01      AUD   
50633   9/16/2015 14:21      failed        1      ES  1.131460e+00      EUR   
63275    9/2/2016 20:04      failed        1      US  0.000000e+00      USD   
40169    2/4/2015 21:47      failed        1      US  1.000000e+00      USD   
6198   10/11/2011 23:14      failed        1      US  1.000000e+00      USD   

         pledged  average_per_backer  
12253  8600000.0          135.612464  
62904  6470000.0           41.761873  
26917  6230000.0          341.931943  
45290  5550000.0           85.559684  
46419  4080000.0          195.075305  
...          ...                 ...  
43036        1.0            1.000000  
50633        1.0            1.000000  
63275        1.0            1.000000  
40169        1.0            1.000000  
6198         1.0            1.000000  

[56136 rows x 14 columns]

8. Create a DataFrame whose index is the first 10 letters of the alphabet and contains two more columns with first 10 prime numbers and the first 10 fibonacci numbers.


In [8]:
letters = ['a','b','c','d','e','f','g','h','i','j']
prime_nums = [1,3,5,7,9,11,13,17,19,23]
fib_nums = [1,1,2,3,5,8,13,21,34,55]
df = pd.DataFrame({'prime_num':prime_nums,'fib_num':fib_nums},index=letters)
df

prime_num  fib_num
a          1        1
b          3        1
c          5        2
d          7        3
e          9        5
f         11        8
g         13       13
h         17       21
i         19       34
j         23       55

9. Merge two DataFrames into a single DataFrame: nycflights13.

Please use the following dataset for merging purposes: airlines & flights

In [9]:
airlines = pd.read_csv('datasets/airlines.csv')
flights = pd.read_csv('datasets/flights.csv')

In [10]:
airlines.shape
airlines.head()

carrier                    name
0      9E       Endeavor Air Inc.
1      AA  American Airlines Inc.
2      AS    Alaska Airlines Inc.
3      B6         JetBlue Airways
4      DL    Delta Air Lines Inc.

In [11]:
flights.shape
flights.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute            time_hour  
0      1400     5      15  2013-01-01 05:00:00  
1      1416     5      29  2013-01-01 05:00:00  
2      1089     5      40  2013-01-01 05:00:00  
3      1576     5      45  2013-01-01 05:00:00  
4       762     6       0  2013-01-01 06:00:00

The common column is carrier. I will merge airlines into flights because flights is bigger.


In [12]:
nycflights13 = flights.merge(airlines,on='carrier')
old = nycflights13.shape[0]

10. Are there any missing values in this DataFrame? If so, drop them.

In [13]:
nycflights13.dropna(inplace=True)
new = nycflights13.shape[0]

In [14]:
datadiff = old-new
print(datadiff)

9430


9430 rows were removed because the contain missing values

11. Merge these two DataFrames into a single DataFrame called census.

Please use the following dataset for merging purposes: census-divisions & state-populations

In [15]:
census = pd.read_csv('datasets/census-divisions.csv')
print(census.columns)
state = pd.read_csv('datasets/state-populations.csv')
print(state.columns)
census = census.merge(state)
print(census.columns)

Index(['state', 'region', 'division'], dtype='object')
Index(['state', '2010', '2011', '2012', '2013', '2014', '2015', '2016'], dtype='object')
Index(['state', 'region', 'division', '2010', '2011', '2012', '2013', '2014',
       '2015', '2016'],
      dtype='object')


12. Re-shape census so that one column contains all population measures, and another the year attributes.

In [16]:
re_census = census.melt(id_vars=['state','region','division'],var_name='year',value_name='population')
re_census.head()

state     region     division  year  population
0    Connecticut  Northeast  New England  2010     3579899
1          Maine  Northeast  New England  2010     1327730
2  Massachusetts  Northeast  New England  2010     6565524
3  New Hampshire  Northeast  New England  2010     1316872
4   Rhode Island  Northeast  New England  2010     1053337

13. Group the data by year and summarize it.

In [17]:
re_census.groupby(by='year').describe()

population                                                              \
          count          mean           std       min        25%        50%   
year                                                                          
2010       51.0  6.065651e+06  6.839536e+06  564513.0  1700530.5  4348662.0   
2011       51.0  6.111046e+06  6.907730e+06  567725.0  1713213.0  4369354.0   
2012       51.0  6.156831e+06  6.974942e+06  576765.0  1725818.0  4384799.0   
2013       51.0  6.200096e+06  7.038774e+06  582684.0  1732621.0  4400477.0   
2014       51.0  6.246342e+06  7.110811e+06  583642.0  1741023.0  4413057.0   
2015       51.0  6.292091e+06  7.181603e+06  586555.0  1746940.5  4424611.0   
2016       51.0  6.335834e+06  7.243849e+06  585501.0  1757121.0  4436974.0   

                             
            75%         max  
year                         
2010  6654375.0  37332685.0  
2011  6717221.5  37676861.0  
2012  6776617.0  38011074.0  
2013  6837396.0  38335203.0  
2014  6902053.5  38680810.0  
2015  6988927.5  38993940.0  
2016  7109535.5  39250017.0

14. Group the data by region, division and year and summarize it.


In [18]:
re_census.groupby(by=['region','division','year']).describe()

population                            \
                                     count        mean           std   
region  division           year                                        
Midwest East North Central 2010        5.0   9288169.4  3.115322e+06   
                           2011        5.0   9301433.8  3.109807e+06   
                           2012        5.0   9314559.0  3.104953e+06   
                           2013        5.0   9332093.0  3.099620e+06   
                           2014        5.0   9346265.8  3.091076e+06   
...                                    ...         ...           ...   
West    Pacific            2012        5.0  10185665.0  1.574263e+07   
                           2013        5.0  10274464.0  1.587648e+07   
                           2014        5.0  10371286.2  1.601909e+07   
                           2015        5.0  10468346.0  1.614496e+07   
                           2016        5.0  10560386.6  1.624360e+07   

                                                                              \
                                       min        25%        50%         75%   
region  division           year                                                
Midwest East North Central 2010  5690263.0  6490528.0  9877495.0  11540983.0   
                           2011  5709640.0  6516480.0  9876213.0  11544824.0   
                           2012  5726177.0  6537743.0  9887238.0  11550839.0   
                           2013  5742854.0  6569102.0  9898982.0  11570022.0   
                           2014  5758377.0  6595233.0  9915767.0  11594408.0   
...                                    ...        ...        ...         ...   
West    Pacific            2012   731089.0  1391820.0  3899116.0   6895226.0   
                           2013   736879.0  1406481.0  3925751.0   6968006.0   
                           2014   736705.0  1416349.0  3968371.0   7054196.0   
                           2015   737709.0  1425157.0  4024634.0   7160290.0   
                           2016   741894.0  1428557.0  4093465.0   7288000.0   

                                             
                                        max  
region  division           year              
Midwest East North Central 2010  12841578.0  
                           2011  12860012.0  
                           2012  12870798.0  
                           2013  12879505.0  
                           2014  12867544.0  
...                                     ...  
West    Pacific            2012  38011074.0  
                           2013  38335203.0  
                           2014  38680810.0  
                           2015  38993940.0  
                           2016  39250017.0  

[63 rows x 8 columns]

15. Create a crosstab between state and category showing amount pledged as the value. (ksproject)

In [19]:
table2 = pd.crosstab(data.state, data.category, values=data.usd_pledged, aggfunc='sum').round(0)
table2

category    3D Printing  Academic  Accessories    Action  Animals  Animation  \
state                                                                          
canceled        99265.0    4152.0     121964.0   13414.0      0.0    57440.0   
failed         318220.0   79896.0     261762.0   81821.0  14026.0   465548.0   
live             4541.0      13.0       7904.0       1.0      NaN      494.0   
successful    8093788.0  345303.0    1995300.0  218054.0  93671.0  2490226.0   
suspended       11576.0     253.0       3774.0       NaN      0.0        0.0   
undefined           0.0       NaN          0.0       NaN      NaN        NaN   

category    Anthologies    Apparel      Apps  Architecture  ...  Wearables  \
state                                                       ...              
canceled         4582.0    90389.0  114503.0       12412.0  ...   212609.0   
failed          15346.0   360408.0  418306.0      146565.0  ...   533602.0   
live              572.0    51390.0    7984.0         397.0  ...    11170.0   
successful     767299.0  5376537.0  608410.0      320563.0  ...  5815185.0   
suspended           NaN    56213.0   12990.0         131.0  ...     7801.0   
undefined           NaN        NaN       NaN           NaN  ...        NaN   

category    Weaving       Web  Webcomics  Webseries  Woodworking  Workshops  \
state                                                                         
canceled      158.0   88769.0     5617.0    54193.0        153.0        NaN   
failed       6742.0  320389.0    26734.0   938842.0      34673.0     2364.0   
live            NaN    2525.0    39688.0        0.0        380.0        0.0   
successful   5634.0  877822.0  1087223.0  3758931.0     219834.0    16908.0   
suspended       0.0       0.0        NaN    14878.0         80.0        0.0   
undefined       NaN       NaN        NaN        NaN          NaN        NaN   

category    World Music  Young Adult    Zines  
state                                          
canceled        15143.0        713.0   6940.0  
failed         121625.0      18192.0  17444.0  
live              283.0        266.0     80.0  
successful     983517.0     111885.0  43615.0  
suspended          37.0          NaN      NaN  
undefined           NaN          NaN      NaN  

[6 rows x 158 columns]